Make layer_norm dispatch from yaml file to fix XLA test

xiaomengy opened the pull request:
Summary: Make layer_norm dispatch from yaml file to fix XLA test

Test Plan: buck test mode/dev-nosan caffe2/test:nn -- "LayerNorm"

Differential Revision: D17939919

BIT-silence commented:
link to https://github.com/pytorch/pytorch/issues/27633

ailzhang commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -26,12 +26,51 @@ std::tuple<Tensor, Tensor, Tensor> native_layer_norm(
 }
 
-std::tuple<Tensor, Tensor, Tensor> native_layer_norm_backward(
+std::tuple<Tensor, Tensor, Tensor> layer_norm_cuda(
```
I'm not sure if this is applicable everywhere, but in general we put `_cpu` functions in `native/cpu` folder and `_cuda` functions in `native/cuda` folder to make the structure more clear. Would you mind moving them a bit? 
There's one minor implication that for the functions defined outside `cpu/` and `cuda/` folder they are mostly called through `TypeDefault` and it should call into `at::` functions, but inside `cpu/` and `cuda` we are free to call "local" `at::native::` functions to speed things up.

BIT-silence replied:
I am a little confused here. In my understanding as the https://github.com/pytorch/pytorch/blob/dfa48f9942e95b13875c4b8841e1a3c01f3638a6/aten/src/ATen/native/DispatchStub.h mentioned. native/cpu and native/cuda folders are mainly for kernels, and in native folder we defines dispatches and call them, that what we did here. Another example is Activation.cpp
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.cpp#L343

ailzhang replied:
Yea nvm, it turns out we don't have this separation at all. :P 
But it actually worth noticing that in `_cpu` and `_cuda` functions we can freely call `at::native::` functions and that doesn't break XLA, but in other non-dispatch functions we can only call `at::` functions. 
I've seen people confused about which to use in the past. Maybe we can think about how to organize these better? cc: @gchanan @dzhulgakov 
But that's not orthogonal to this PR for sure. Thanks for reminding!

ailzhang reviewed (commented):
Thanks @BIT-silence!! I only have a minor comment inlined in the PR.

xiaomengy committed:
Make layer_norm dispatch from yaml file to fix XLA test (#28051)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/28051

Make layer_norm dispatch from yaml file to fix XLA test

Test Plan: buck test mode/dev-nosan caffe2/test:nn -- "LayerNorm"

Reviewed By: houseroad

Differential Revision: D17939919

fbshipit-source-id: 38c50c96b4494b727d8ed90d983e3a150d0e5b8d
